hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsOutputStream.java

```diff
@@ -309,15 +316,23 @@ private synchronized void flushWrittenBytesToService(boolean isClose) throws IOE
             throw new FileNotFoundException(ex.getMessage());
           }
         }
-
         if (ex.getCause() instanceof AzureBlobFileSystemException) {
           ex = (AzureBlobFileSystemException) ex.getCause();
         }
         lastError = new IOException(ex);
         throw lastError;
       }
     }
-    flushWrittenBytesToServiceInternal(position, false, isClose);
+    shrinkWriteOperationQueue();
```

Comment:
AbfsOutputStream.close() call flow is:
close() -> flushInternal() -> flushWrittenBytesToService() **-> flushWrittenBytesToServiceInternal()**
Within flushWrittenBytesToServiceInternal() is where service "Flush" gets called. 

Above change removes call to flushWrittenBytesToServiceInternal() and replaces it with shrinkWriteOperationQueue().  This is going to prevent flush for AbfsOutputStream.close() calls. 

This might be ok for AppendBlob, but BlockBlob behaviour needs to be retained where AbfsClient->Flush() will get triggered.

Reply:
as discussed offline, please look at new REST API. the code is same for both the appendBlob and BlockBlob. THe difference at the OutputStream Code is to use a single upload thread for appendBlob and multiple upload threads for block Blob. Its a lot of code rewrite, hence the full file read might be more useful.